src/dist/mod.rs

```diff
@@ -30,6 +32,124 @@ pub mod http;
 #[cfg(test)]
 mod test;
 
+// TODO: paths (particularly outputs, which are accessed by an unsandboxed program)
+// should be some pre-sanitised AbsPath type
+
+pub use self::path_transform::PathTransformer;
+
+#[cfg(target_os = "windows")]
+mod path_transform {
+    use std::collections::HashMap;
+    use std::path::{Component, Path, PathBuf, Prefix};
+    use std::str;
+
+    pub struct PathTransformer {
+        dist_to_local_path: HashMap<String, PathBuf>,
+    }
+
+    impl PathTransformer {
+        pub fn new() -> Self {
+            PathTransformer {
+                dist_to_local_path: HashMap::new(),
+            }
+        }
+        pub fn to_dist_abs(&mut self, p: &Path) -> Option<String> {
+            if !p.is_absolute() { panic!("non absolute path {:?}", p) }
+            self.to_dist(p)
+        }
+        pub fn to_dist(&mut self, p: &Path) -> Option<String> {
+            let mut components = p.components();
+
+            let maybe_dist_prefix = if p.is_absolute() {
+                let prefix = components.next().unwrap();
+                let dist_prefix = match prefix {
+                    Component::Prefix(pc) => {
+                        match pc.kind() {
+                            // Transforming these to the same place means these may flip-flop
+                            // in the tracking map, but they're equivalent so not really an
+                            // issue
+                            Prefix::Disk(diskchar) |
+                            Prefix::VerbatimDisk(diskchar) => {
+                                assert!(diskchar.is_ascii_alphabetic());
+                                format!("disk-{}", str::from_utf8(&[diskchar]).unwrap())
+                            },
+                            Prefix::Verbatim(_) |
+                            Prefix::VerbatimUNC(_, _) |
+                            Prefix::DeviceNS(_) |
+                            Prefix::UNC(_, _) => return None,
+                        }
+                    },
+                    _ => panic!("unrecognised start to path {:?}", p),
+                };
+
+                let root = components.next().unwrap();
+                if root != Component::RootDir { panic!("unexpected non-root component in {:?}", p) }
+
+                Some(dist_prefix)
+            } else {
+                None
+            };
+
+            let mut dist_suffix = String::new();
+            for component in components {
+                let part = match component {
+                    Component::Prefix(_) |
+                    Component::RootDir => panic!("unexpected part in path {:?}", p),
+                    Component::Normal(osstr) => osstr.to_str()?,
+                    // TODO: should be forbidden
```

Comment:
Did you run into cases where these were necessary in practice? If not, it seems like it'd be better to forbid them now before anyone relies on this working.